# CSE 25 - Music Recommendation System

**Group Members:** Ovi Joshi, Jackson Horman, Johanna Sanchez

**Professor:** Minnes


## Project Summary

This project builds a content-based music recommendation system trained on Spotify audio features. 
Rather than relying on user listening history or popularity, our system analyzes the actual sound 
characteristics of a song such as danceability, energy, tempo, loudness, and valence to find 
musically similar tracks.

We train a supervised neural network on these audio features to classify songs by genre. 
Once trained, we extract the model's internal learned representations (embeddings) and use 
cosine similarity to compare songs and return the most similar recommendations.

The dataset consists of over 114,000 songs across 125 genres sourced from Kaggle's Spotify 
Tracks Dataset. We will evaluate our model using accuracy, precision, recall, and F1-score, 
and compare it against a Logistic Regression baseline.

## 1: Importing Packages

In [6]:
# core libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

print("All packages imported successfully")

All packages imported successfully


## 2: Load the Data
We define a 'load_data' function that reads the Spotify dataset from a CSV file and prepares it for use in our model. The function does three things:


-Reads the CSV file into a pandas dataframe

-Drops any rows with missing values in our audio feature columns

-Removes duplicate songs so we don't train on the same song twice


In [7]:
def load_data(filepath):
    # Read the CSV file into a pandas dataframe
    df = pd.read_csv(filepath)

    # These are the audio features we will use to train our model
    feature_cols = [
        'danceability',     # How suitable a track is for dancing
        'energy',           # Intensity and activity of the track
        'loudness',         # Overall loudness in decibels
        'speechiness',      # Presence of spoken words
        'acousticness',     # How acoustic the track is
        'instrumentalness', # Whether a track has no vocals
        'liveness',         # Presence of a live audience
        'valence',          # Musical positiveness of the track
        'tempo'             # Speed of the track in BPM
    ]

    # Remove rows where any feature value or genre is missing
    df = df.dropna(subset=feature_cols + ['track_genre'])

    # Remove duplicate songs keeping only the first occurrence
    df = df.drop_duplicates(subset='track_id')

    # Print a summary so we can verify the data loaded correctly
    print(f"Dataset shape: {df.shape}")
    print(f"Number of genres: {df['track_genre'].nunique()}")
    print(f"Sample:\n{df[feature_cols].head()}")

    return df, feature_cols

# Call the function with the path to our dataset
df, feature_cols = load_data('data/dataset.csv')

Dataset shape: (89741, 21)
Number of genres: 113
Sample:
   danceability  energy  loudness  speechiness  acousticness  \
0         0.676  0.4610    -6.746       0.1430        0.0322   
1         0.420  0.1660   -17.235       0.0763        0.9240   
2         0.438  0.3590    -9.734       0.0557        0.2100   
3         0.266  0.0596   -18.515       0.0363        0.9050   
4         0.618  0.4430    -9.681       0.0526        0.4690   

   instrumentalness  liveness  valence    tempo  
0          0.000001    0.3580    0.715   87.917  
1          0.000006    0.1010    0.267   77.489  
2          0.000000    0.1170    0.120   76.332  
3          0.000071    0.1320    0.143  181.740  
4          0.000000    0.0829    0.167  119.949  


In [8]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, LabelEncoder

def preprocess_data(df,feature_cols):
    X = df[feature_cols]
    y = df['track_genre']

    scaler = StandardScaler()
    X_scale = scaler.fit_transform(X)
    encoder = LabelEncoder()
    y_int = LabelEncoder().fit_transform(y)

    X_train, X_test, y_train, y_test = train_test_split(
        X_scale,y_int,test_size=0.2,
        train_size = 0.8,
        random_state = 25,
        shuffle = True,
        stratify = y_int
        )
    
    print("Preprocessing complete")
    print("Train shape:", X_train.shape)
    print("Test shape:", X_test.shape)

    return X_train, X_test, y_train, y_test, StandardScaler(), LabelEncoder()

X_train, X_test, y_train, y_test, scalar, encoder = preprocess_data(df, feature_cols)


Preprocessing complete
Train shape: (71792, 9)
Test shape: (17949, 9)


## 3: Handling Class Imbalance

Since some genres have more songs than others, we compute class weights so the model pays equal attention to all genres during training. The genres with fewer songs will get a higher weight so the model does not ignore them.

In [9]:
from sklearn.utils.class_weight import compute_class_weight

class_weights = compute_class_weight(
    class_weight='balanced',
    classes=np.unique(y_train),
    y=y_train
)

print("Class weights computed successfully")
print(f"Number of classes: {len(class_weights)}")
print(f"Min weight: {class_weights.min():.4f}")
print(f"Max weight: {class_weights.max():.4f}")

Class weights computed successfully
Number of classes: 113
Min weight: 0.7942
Max weight: 10.7683


## 4: Building the Neural Network

We define a neural network using TensorFlow and Keras. The model will take in the 9 audio features as input and will learn how to classify songs by genre. 

Our neural network also has 3 layers:

-Input layer: 64 neurons that take in the 9 audio features

-Hidden layer: 32 neurons that learn compressed representations of songs(embeddings) 

-Output layer: 113 neurons, one for each genre, that predicts that genre of a song


The hidden layer will help us extract 32 neurons embeddings and use them to compare songs using cosine similarity

In [12]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Input

input_size = X_train.shape[1]
num_classes = len(np.unique(y_train))

# Define the neural network
model = Sequential()

# Input layer
model.add(Input(shape=(input_size,)))

# First layer
model.add(Dense(64, activation='relu'))

# Hidden layer 
model.add(Dense(32, activation='relu'))

# Output layer 
model.add(Dense(num_classes, activation='softmax'))

# Compile with class weights loss function
model.compile(
    optimizer='adam',
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)

model.summary()

Model: "sequential_3"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ dense_6 (Dense)                      │ (None, 64)                  │             640 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_7 (Dense)                      │ (None, 32)                  │           2,080 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_8 (Dense)                      │ (None, 113)                 │           3,729 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 6,449 (25.19 KB)

 Trainable params: 6,449 (25.19 KB)

 Non-trainable params: 0 (0.00 B)

## Model Summary Explanation

The table above shows the structure of our neural network. Each song starts 
as 9 audio features and passes through 3 layers. The first layer expands it 
to 64 numbers, the second layer compresses it down to 32 numbers which are 
our embeddings, and the final layer outputs 113 probabilities(one for each 
genre). The genre with the highest probability is the model's prediction. 
In total the model has 6,449 learnable values that get updated during training.

## 5: Extracting Embeddings

next step ...